### PR TITLE
Fix issue where structure tag could get stuck at the wrong location

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Partial fix of https://github.com/dotnet/roslyn/issues/73799

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@olegtk @AmadeusW @veler can you point me to where your code for IStructureTag structure tags is?  I have a feelling you are not using .EdgeInclusive as your span tracking mode for them, causing them to not move to the desired location as the user edits at the start of a structure tag.  

If that's the case, we'll likely want you to change that to fix https://github.com/dotnet/roslyn/issues/73799 completely.

**Comment by ToddGrun** [comment]
> can you point me to where your code for IStructureTag structure tags is

Is [this](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS-Platform?path=/src/Editor/Text/Impl/Structure/StructureSpanningTree/StructureMappingUtil.cs&version=GBmain&line=227&lineEnd=228&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) it?

**Comment by CyrusNajmabadi** [comment]
Thanks.  Given:

```
            guideLineSpanStart = guideLineSpanStart != null ?
                MappingPointSnapshot.Create(
                    targetSnapshot,
                    new SnapshotPoint(tag.Snapshot, guideLineSpanStart.Value),
                    PointTrackingMode.Positive, null).GetPoint(targetSnapshot, PositionAffinity.Predecessor) :
                    null;
```

This should likely move to PointTrackingMode.Negative @olegtk

File: src/EditorFeatures/Core/Structure/AbstractStructureTaggerProvider.cs [hunk 1]
```diff
@@ -219,11 +218,33 @@ private void ProcessSpans(
...
+        if (latestTag is not StructureTag latestStructureTag || previousTag is not StructureTag previousStructureTag)
+        {
+            Contract.Fail("Tags were the wrong type");
+            return latestTag == previousTag;
+        }
+
+        var latestSnapshot = latestStructureTag.Snapshot;
+        var previousSnapshot = previousStructureTag.Snapshot;
+
+        var previousStructureStart = new SnapshotPoint(previousSnapshot, previousStructureTag.HeaderSpan.Start);
```
Comment (ToddGrun) [inline src/EditorFeatures/Core/Structure/AbstractStructureTaggerProvider.cs:232]: >HeaderSpan [](http://example.com/codeflow?start=94&length=10)

Is this just about the guideline behavior? If so, would GuideLineSpan.Start be appropriate to use? FYI, I believe you could also use a SnapshotSpan and SpanTrackingMode.Inclusive

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Structure/AbstractStructureTaggerProvider.cs:232]: that seems reasonable.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Structure/AbstractStructureTaggerProvider.cs:232]: We can't use guidelinespan as we don't set it.  it's just:

```c#
 public Span? GuideLineSpan => null;
```

we defer to the editor to determine that span.

**Comment by CyrusNajmabadi** [comment]
@ToddGrun ptal.

**Comment by CyrusNajmabadi** [comment]
@JoeRobich @akhera99 ptal.

File: src/EditorFeatures/Core/Structure/AbstractStructureTaggerProvider.cs (hunk 1 above)
Comment (ToddGrun) [reply src/EditorFeatures/Core/Structure/AbstractStructureTaggerProvider.cs:232]: I still feel like mapping a span is better than two calls to mapping points.

var previousStructureStart = new SnapshotSpan(previousSnapshot, previousStructureTag.HeaderSpan.Start, 0);

if (0 <previousStructureStart.TranslateTo(latestSnapshot, PointTrackingMode.Negative).Length)

Comment (ToddGrun) [reply src/EditorFeatures/Core/Structure/AbstractStructureTaggerProvider.cs:232]: To clarify, I was thinking of something like the following (with the comment needing to be altered)

```
    protected sealed override bool TagEquals(IContainerStructureTag latestTag, IContainerStructureTag previousTag)
    {
        if (latestTag is not StructureTag latestStructureTag || previousTag is not StructureTag previousStructureTag)
        {
            Contract.Fail("Tags were the wrong type");
            return latestTag == previousTag;
        }

        // If these tags are the same object in memory, then they are obviously equal
        if (latestTag == previousTag)
            return true;

        var latestSnapshot = latestStructureTag.Snapshot;
        var previousSnapshot = previousStructureTag.Snapshot;

        var previousStructureStart = new SnapshotSpan(previousSnapshot, previousStructureTag.HeaderSpan.Start, 0);

        if (previousStructureStart.TranslateTo(latestSnapshot, SpanTrackingMode.EdgeInclusive).IsEmpty)
        {
            // We can't know that how we think this block moved is actually how the editor actually moved it.
            // Specifically, the tracking mode is an implementation detail.  As such, we don't want to reuse this tag as
            // its stale data (as mapped by the editor) may not be where we'd expect the new block's data to be.  This
            // can happen when the user types right at the start of a structure tag, causing it to move inwards
            // undesirably.

            return latestTag.Equals(previousTag);
        }

        return false;
    }
```

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Comment by CyrusNajmabadi** [comment]
Tlaked offline.  Going to take the code as is as i personally think it is clearer, and this is not a hot path.

**Pull request merged**